hack/test-e2e.sh

```diff
@@ -0,0 +1,67 @@
+#!/usr/bin/env bash
+
+# Copyright 2021 The KCP Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+set -o nounset
+set -o pipefail
+set -o xtrace
+set -o errexit
+
+export PROMETHEUS_URL=""
+USE_PROMETHEUS="${USE_PROMETHEUS:-}"
+if [ -n "${USE_PROMETHEUS}" ]; then
+  touch .prometheus-config.yaml
+  ./hack/tools/prometheus --storage.tsdb.path=".prometheus_data" --config.file=.prometheus-config.yaml --web.enable-lifecycle &
+  PROM_PID=$!
+  export PROMETHEUS_URL="http://localhost:9090"
+  while ! curl -s http://localhost:9090/-/ready; do
+    sleep 1
+  done
+fi
+
+if [ -n "${TEST_SERVER:=}" ]; then
+  ${TEST_SERVER} 2>&1 &
+  TEST_SERVER_PID=$!
+  echo "test server PID ${TEST_SERVER_PID}"
+  while [ ! -f "${WORK_DIR}/.kcp/ready-to-test" ]; do
+    sleep 1
+  done
+fi
+
+set +o errexit
+${TEST_CMD}
+TEST_EXIT_CODE=$?
+set -o errexit
+
+echo "Test terminated with ${TEST_EXIT_CODE}"
+
+if [ -n "${TEST_SERVER}" ]; then
+  echo "killing ${TEST_SERVER_PID}"
+  kill -TERM "${TEST_SERVER_PID}"
+fi
+
+if [ -n "${WAIT:=}" ]; then
+  echo "waiting for ${TEST_SERVER_PID}"
+  wait "${TEST_SERVER_PID}"
+fi
+
+if [ -n "${USE_PROMETHEUS}" ] && [ -n "${ARTIFACT_DIR:=}" ]; then
+  kill -TERM ${PROM_PID}
+  while [ -f .prometheus_data/lock ]; do sleep 1; done
```

Comment:
General comment: anyplace we are waiting for something to happen, please echo before ("waiting for x") and after.

Reply:
fixed